### ISSUE TITLE
Upgrade lost settings, import caused breakage afterwards

### ISSUE THREAD

**Issue by pjeby** [crittermike/shortkeys#598]
Earlier this evening I saw a message on one of my computers about how Shortkeys is now running in the background and at system start.  So I went to use one of its function keys to make sure everything was still working, and it didn't work.  So I went to options and found my configuration empty - the upgrade deleted the configuration, it seems!

I went to a different computer -- same thing.  Went to a laptop, pulled the network cable and woke it from sleep so I could export the settings from that machine before that copy's settings could be deleted, too.

I then imported the settings back onto other computers, but couldn't get anything to work.  After a while, I noticed that there was an empty shortcut present above the imported items.  After deleting the empty item and saving, I got things working again, at least on my main machine.  (Haven't tested the others yet.)

tl;dr: Apparently, importing when you don't have any shortcuts (due to the upgrade deleting them) results in a blank shortcut above the imported shortcuts, which then appears to keep the rest from working correctly.

**Comment by kouki-itou** [comment]
Same here.
and I noticed CSP error happened.

**Comment by nerdgirl** [comment]
Same here. Thankfully I had exported my settings the other day so at least I have the config. But nothing is working even after reimporting them.

**Comment by crittermike** [comment]
Will publish a revert while I investigate!

**Comment by crittermike** [comment]
I just submitted a revert back to 4.0.2 but I have to wait for the webstore to review/publish it. All of your settings should still be there once that is published, as long as you didn't click submit on the settings form.

In the meantime, I'll investigate the issue.

**Comment by w4tchdoge** [comment]
> [...] we changed configs storage for the keys from local storage [...]

Thank you _so much_ for pointing this out, I was able to go check the local storage using DevTools and recover my configuration. Saves me the hassle of re-writing all the JavaScript (which I had also not saved anywhere) for the shortkeys I had which were just running those in the browser console.

**Comment by crittermike** [comment]
No worries @fastndead! You're a champ - this is on me for somehow missing it in my regression testing.

**Comment by malkomalko** [comment]
Just wanted to add that if you go to the extensions options page and open up dev tools you can get your old data by looking at `localStorage.shortkeys`

**Comment by malkomalko** [comment]
Upon looking at the code I think it's because `this.keys = this.keys.concat(JSON.parse(this.importJson)),` inside `importKeys` should be `this.keys = this.keys.concat(JSON.parse(this.importJson).keys),` if I'm understanding this correctly.

**Comment by oniatsu** [comment]
I note here because I saw a way to port the previous localStorage to read chrome.storage.
[chrome.storage  |  API  |  Chrome for Developers](https://developer.chrome.com/docs/extensions/reference/api/storage)

**Comment by webaschtl** [comment]
Today my javascripts of Shortkey are also lost. After importing them again (yes, I have a backup ;-)), they do not work. And I get this message:

`Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'toLowerCase')
    at Object.activateKey (content.js:1:6082)
    at content.js:1:6484
    at Array.forEach (<anonymous>)
    at content.js:1:6469`

Has this something to do with this "bug"?

**Comment by pjeby** [comment]
Check if you have a blank entry in your list of shortcuts, and if so, delete it.  I don't know if that's the problem you're having, as I never checked the console, but if there's a blank shortcut in the list then none of them will work.  And if you imported after the upgrade and never deleted the blank shortcut created by the upgrade, then you probably need to delete it.

**Comment by crittermike** [comment]
Hi all, quick update, the reverted package keeps getting rejected by the webstore team for annoying reasons. I just submitted the 3rd attempt - hopefully that one goes through.

**Comment by webaschtl** [comment]
> Check if you have a blank entry in your list of shortcuts, and if so, delete it. I don't know if that's the problem you're having, as I never checked the console, but if there's a blank shortcut in the list then none of them will work. And if you imported after the upgrade and never deleted the blank shortcut created by the upgrade, then you probably need to delete it.

Hm, I still do not know the cause. There was no blank entry. 
On my second PC the shortcuts had also gone. But with the tip in this thread (https://github.com/crittermike/shortkeys/issues/598#issuecomment-1944845929) to restore the "old" data, the Shortkeys now working again.

**Comment by dannypernik** [comment]
> Just wanted to add that if you go to the extensions options page and open up dev tools you can get your old data by running `copy(JSON.stringify(JSON.parse(localStorage.shortkeys).keys, null, 2))` which will copy the config to your clipboard. Then you should be able to paste this into import to get back up.

Forgive my naivety, but where do I run this? I tried the developer console but got a syntax error.

@crittermike any update from the webstore?

**Comment by crittermike** [comment]
@dannypernik the revert should be posted now - 4.1.2. I've heard from others that that fixed things for them. No luck from your end?

**Comment by dannypernik** [comment]
@crittermike I confirmed that version 4.1.2 is active, but former shortcuts are not displaying. They weren't the most sophisticated or numerous shortcuts (yet) so I'm fine to recreate them if necessary.

@hababr here's what I'm seeing:
![image](https://github.com/crittermike/shortkeys/assets/3347095/3c421bc4-e761-4f71-9caf-1458774af348)

**Comment by dannypernik** [comment]
@crittermike following up here. I added a few of my shortcuts back to my blank options but never got my old ones back.
